feat(SPV-1442) : Pagnation code handled by backend queries

cain4chain opened the pull request:
feat(pagination): implement server-side pagination for contacts
feat(pagination): add server-side pagination for paymails
feat(pagination): implement server-side pagination for access keys
feat(pagination): add server-side pagination for transactions
feat(pagination): implement server-side pagination for xpubs
feat(hooks): create usePagination hook for centralized pagination logic
feat(hooks): implement useTable hook for table state management
feat(utils): add query options with pagination support for API calls
fix(pagination): ensure consistent page index conversion between UI and API
feat(components): add pagination support to mobile data tables

# Pull Request Checklist

- [x] 📖 I created my PR using provided  : [CODE_STANDARDS](https://github.com/bitcoin-sv/spv-wallet-admin/blob/main/.github/CODE_STANDARDS.md)
- [x] 📖 I have read the short Code of Conduct: [CODE_OF_CONDUCT](https://github.com/bitcoin-sv/spv-wallet-admin/blob/main/.github/CODE_OF_CONDUCT.md)
- [x] 🏠 I tested my changes locally.
- [ ] ✅ I have provided tests for my changes.
- [x] 📝 I have used conventional commits.
- [ ] 📗 I have updated any related documentation.
- [x] 💾 PR was issued based on the Github or Jira issue.

cain4chain committed:
feat(SPV-1442) : Pagnation code handled by backend queries

feat(pagination): implement server-side pagination for contacts
feat(pagination): add server-side pagination for paymails
feat(pagination): implement server-side pagination for access keys
feat(pagination): add server-side pagination for transactions
feat(pagination): implement server-side pagination for xpubs
feat(hooks): create usePagination hook for centralized pagination logic
feat(hooks): implement useTable hook for table state management
feat(utils): add query options with pagination support for API calls
fix(pagination): ensure consistent page index conversion between UI and API
feat(components): add pagination support to mobile data tables

chris-4chain commented on src/components/AccessKeysTabContent/AccessKeysTabContent.tsx:
```diff
@@ -36,10 +38,12 @@ export const AccessKeysTabContent = ({ accessKeys, hasRevokeKeyDialog }: AccessK
             </div>
             <div className="sm:hidden">
-              <AccessKeysMobileList accessKeys={accessKeys} />
+              <AccessKeysMobileList accessKeys={accessKeys} pagination={pagination} />
```
Maybe it is for the future considerations, but I think, switching between mobile and normal version should not be done by only "hiding" components. 

React is mounting and rendering components for both versions (even for css-hided) which is not good from the performance perspective, and it could cause weird behaviour, like both components try to do some "side-effect" or load data on mount.

I would suggest to employ such hook as `useMediaQuery`

chris-4chain commented on src/components/DataTable/useTable.ts:
```diff
@@ -0,0 +1,86 @@
+    getPaginationRowModel: getPaginationRowModel(),
+    getSortedRowModel: getSortedRowModel(),
+    initialState: {
+      sorting: initialSorting ? initialSorting : defaultInitialSorting,
```
```suggestion
sorting: initialSorting ?? defaultInitialSorting,
```

chris-4chain commented on src/components/DataTable/useTable.ts:
```diff
@@ -0,0 +1,86 @@
+            pageIndex: pagination.currentPage,
+            pageSize: pagination.pageSize,
+          }
+        : {
+            pageIndex: 0,
+            pageSize: 10,
+          },
```
Consider defining `defaultInitialPagination`, similarly to `defaultInitialSorting`.

cain4chain replied:
Thanks, I will update and provide for some common pagination values

chris-4chain commented on src/components/DataTable/useTable.ts:
```diff
@@ -0,0 +1,86 @@
+  columns,
+  data,
+  initialSorting,
+  pagination,
```
I'm wondering if the `manualPagination` as an explicit prop is necessary. 
Can't you just deduce, that:
```js
const manualPagination = !!pagination
```

cain4chain replied:
yes - will remove from components, thanks!

chris-4chain commented on src/routes/admin/_admin.paymails.tsx:
```diff
@@ -13,11 +13,24 @@ import {
+import { PaymailAddress } from '@bsv/spv-wallet-js-client';
+
+// Define interface for API response
+interface PaymailsApiResponse {
+  content: PaymailAddress[];
+  page: {
+    size: number;
+    number: number;
+    totalElements: number;
+    totalPages: number;
+  };
+}
```
Consider use some typescript features to define the `page` part **once**, and reuse it for all the responses (like `PaymailApiResponse`, `ContactApiResponse` etc.)

chris-4chain commented on src/routes/admin/_admin.paymails.tsx:
```diff
@@ -52,36 +75,83 @@ export function Paymails() {
-  }, [tab]);
+  }, [tab, navigate]);
+
+  // Use useCallback to memoize pagination handlers
+  const handlePageChange = useCallback(
+    (newPage: number) => {
+      navigate({
+        search: (old) => ({
+          ...old,
+          page: convertToApiPage(newPage),
+          size: old.size || 10,
+        }),
+        replace: true,
+      });
+    },
+    [navigate],
+  );
+
+  const handlePageSizeChange = useCallback(
+    (newSize: number) => {
+      navigate({
+        search: (old) => ({
+          ...old,
+          size: newSize,
+          page: 1, // Reset to first page when changing page size
+        }),
+        replace: true,
+      });
+    },
+    [navigate],
+  );
```
It would be great to handle this once maybe in `usePagination` hook or `<DataTable/>` component. 

Or provide some custom hook to be easily used by every component. 

I'm aware, that `useSearch` & `useNavigate` can be hard for "generic" functions - but actually, I've already succeeded with similar function - maybe you could implement something alike.

https://github.com/bitcoin-sv/spv-wallet-admin/blob/35bf5be14f3228ee13f7379b6a1200c0589d0c86/src/hooks/useSearchParam.ts#L4-L28

cain4chain replied:
Thanks, this looks good, I will implement the changes

cain4chain committed:
feat(SPV-1442) : Updates as per comments

feat(Pagination) : Update pagination types and constants
feat(usePagination) : Common hook for useRoutePagination

chris-4chain commented on src/components/AccessKeysColumns/AccessKeysColumnsMobile.tsx:
```diff
@@ -55,10 +55,10 @@ export const AccessKeyMobileItem = ({ accessKey }: AccessKeyMobileItemProps) =>
     original: accessKey,
     getValue: (key: string) => (key === 'status' ? accessKey.status : undefined),
-  } as Row<AccessKeysColumns>;
+  } as unknown as Row<AccessKey>;
```
Why is this necessary ? (The `unknown` casting)

chris-4chain commented on src/components/DataTable/DataTable.tsx:
```diff
@@ -13,32 +13,44 @@ import {
+  | ContactExtended
+  | PaymailExtended
+  | XpubExtended
+  | { id?: string; url?: string; status?: string };
```
Consider making a type for it to be consistent and to describe (by type name) what it is.

chris-4chain commented on src/components/DataTable/MobileDataTable.tsx:
```diff
@@ -24,31 +16,34 @@ interface MobileDataTableProps<TData extends RowType, TValue> {
+  useEffect(() => {
+    setExpandedItems([]);
+    setIsAllExpanded(false);
+  }, [table.getState().pagination.pageIndex, table.getState().pagination.pageSize]);
```
You could make a local variable (constant) for the pagination state.

chris-4chain commented on src/components/DataTable/usePagination.ts:
```diff
@@ -0,0 +1,134 @@
+  const total = manualPagination && totalRecords !== undefined ? totalRecords : table.getFilteredRowModel().rows.length;
+
+  // Track the current page size in local state to ensure the UI stays in sync
+  const [currentPageSize, setCurrentPageSize] = useState<number>(table.getState().pagination.pageSize);
```
You could also make a local const variable for `table.getState().pagination`

chris-4chain commented on src/components/DataTable/usePagination.ts:
```diff
@@ -0,0 +1,134 @@
+  // Track the current page size in local state to ensure the UI stays in sync
+  const [currentPageSize, setCurrentPageSize] = useState<number>(table.getState().pagination.pageSize);
+
+  // Initialize currentPageSize on mount and when table changes
+  useEffect(() => {
+    setCurrentPageSize(table.getState().pagination.pageSize);
+  }, [table]);
+
+  // Update local state when table pagination state changes
+  useEffect(() => {
+    // Always update the local state to match the table's page size
+    setCurrentPageSize(table.getState().pagination.pageSize);
+  }, [table.getState().pagination.pageSize]);
+
+  useEffect(() => {
+    setCurrentPageSize(table.getState().pagination.pageSize);
+  }, [table.getState().pagination]);
```
It can be one `useEffect` with those three in one dependency list.

chris-4chain commented on src/components/DataTable/usePagination.ts:
```diff
@@ -0,0 +1,134 @@
+ * @param routeID - The route ID to navigate to
+ * @returns Pagination handlers
+ */
+export function useRoutePagination<TRouter extends AnyRouter = RegisteredRouter, const TFrom extends string = string>(
+  routeID: ConstrainLiteral<TFrom, RouteIds<TRouter['routeTree']>>,
+) {
```
🚀 👍

chris-4chain commented on src/routes/admin/_admin.access-keys.tsx:
```diff
@@ -13,142 +12,214 @@ import {
+
+  const pagination = useRoutePagination('/admin/_admin/access-keys');
+
+  // Local filter state for input field
+  const [filter, setFilter] = useState<string>(xpubId || '');
+  const [debouncedFilter] = useDebounce(filter, 500);
+
+  // Update URL when the debounced filter changes
+  useEffect(() => {
+    navigate({
+      search: (old) => ({
+        ...old,
+        xpubId: debouncedFilter || undefined,
+      }),
+      replace: true,
+    });
+  }, [debouncedFilter, navigate]);
+
+  // Sync filter input when URL xpubId changes
+  useEffect(() => {
+    setFilter(xpubId || '');
+  }, [xpubId]);
```
Since my last PR, the `Search` component does "debouncing". I think, `useSearchParam` hook should be used here.
